core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java

```diff
@@ -134,31 +138,82 @@ public static boolean isMergeJoinSupported(JoinRelType joinType) {
       final RelTraitSet required) {
     // Required collation keys can be subset or superset of merge join keys.
     RelCollation collation = required.getCollation();
+    int leftInputFieldCount = left.getRowType().getFieldCount();
+
     List<Integer> reqKeys = RelCollations.ordinals(collation);
-    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
+    List<Integer> leftKeys = joinInfo.leftKeys.toIntegerList();
+    List<Integer> rightKeys =
+        joinInfo.rightKeys.incr(leftInputFieldCount).toIntegerList();
 
+    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
     ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
     ImmutableBitSet rightKeySet = ImmutableBitSet.of(joinInfo.rightKeys)
-        .shift(left.getRowType().getFieldCount());
+        .shift(leftInputFieldCount);
 
-    // Only consider exact key match for now
     if (reqKeySet.equals(leftKeySet)) {
+      // if sort keys equal to left join keys, we can pass through all collations directly.
       Mappings.TargetMapping mapping = buildMapping(true);
       RelCollation rightCollation = collation.apply(mapping);
       return Pair.of(
           required, ImmutableList.of(required,
           required.replace(rightCollation)));
+    } else if (containsOrderless(leftKeys, collation)) {
```

Comment:
I think it would be helpful to describe the general logic and all the different cases in a javadoc on `passThroughTraits`

Reply:
Got it. Will give a summary as java doc.

Reply:
Added a summary.